HW3 - Interface Implementation

BhanuPrakashNani opened the pull request:
HW3

- Implemented full AIConversationClient interface for both OpenAI and Cerebras APIs with all required methods  
- Added comprehensive session management including creation, tracking and termination  
- Built complete chat history handling with retrieval, summarization and JSON export capabilities  
- Supported model switching between different AI models for each provider  
- Implemented usage metrics tracking for tokens, API calls and cost estimation  
- Created thorough test suite with 100% coverage including interface compliance and error cases  
- Enforced strict typing with mypy and maintained PEP 8 standards throughout  
- Used in-memory session storage with thread-safe operations (production would need persistence layer)  
- Added robust error handling for both API failures and client-side validations  
- Optimized API usage with efficient chat history serialization and minimal requests  
- Included example usage scripts demonstrating core functionality  
- Documented all methods with complete docstrings including args, returns and raises  
- Testing requires setting CEREBRAS_API_KEY environment variables

Commit:
Clear base

Commit:
Modify config.yml

Loke7132 committed:
Base Interface implementation

Loke7132 committed:
test-1

Loke7132 committed:
test-2

Loke7132 committed:
test-3

Loke7132 committed:
test-4

Loke7132 committed:
Remove environment files from repository

Loke7132 committed:
test-5

Loke7132 committed:
test-6

Loke7132 committed:
test-7

Loke7132 committed:
test-8

Commit:
Modified function to satisfy type-check

AdithyahNair commented on .gitignore:
```diff
@@ -75,3 +75,39 @@ ipython_config.py
 
 # uv (dependency manager)
 .uv/
+
+# Environment variables
+.env
+.env.local
+.venv
+env/
+venv/
+ENV/
+env.bak/
+venv.bak/
+
+# VS Code
+.vscode/
+*.code-workspace
+
+# PyCharm
+.idea/
+*.iml
+*.iws
+.idea_modules/
+
+# macOS
+.DS_Store
+.AppleDouble
+.LSOverride
+
+# Windows
```
What is the reason for adding these many extensions to the gitignore file?

KarthikVV24-11-1999 commented:
As suggested by the TAs and the professor, only changes relevant to this particular homework are to be present in this PR
You shall change the base for this PR to "interface-definition" branch to achieve the same

Loke7132 committed:
TA frrdback changes

Loke7132 committed:
feed back changes from HW2

Loke7132 committed:
feed back changes from HW2 change-1

Loke7132 committed:
feed back changes from HW2 change-2

Loke7132 committed:
feed back changes from HW2 change-3

Loke7132 committed:
feed back changes from HW2 change-4

Loke7132 committed:
feed back changes from HW2 change-5

Loke7132 committed:
feed back changes from HW2 change-6

Loke7132 committed:
feed back changes from HW2 change-7

Loke7132 committed:
feed back changes from HW2 change-8

Loke7132 committed:
feed back changes from HW2 change-9

aatish2098 commented:
- Missing docstrings in factory.py and cli.py—please document all public functions and CLI commands to improve API discoverability.
- Unit tests cover core interface and client logic, but add dedicated tests for CLI execution and error scenarios (e.g., invalid session IDs, network failures) to boost coverage.
- Enhance cerebras_client.py with explicit parameter validation and robust error handling 
- Changes not relevant to implementation should not be present in this PR (things like changing test for other components, etc). Merge them separately if you require them in your repository, but having them in this implementation PR will only harm your grade.

AdithyahNair commented:
The codebase is well-structured, thoroughly tested, and follows good software engineering practices. 

- The abstraction with the `AIConversationClient` interface meets the requirements of dependency injection
- Good use of the Factory Design Pattern

- Test suite's thorough and has good coverage
- Effective use of mock for testing

- The doctrings have Args/Returns/Raises sections, which make them very comprehensive and easy to read
-Good example usage in README

- factory.py needs more comprehensive docstrings, however

- I'd suggest creating custom exception classes instead of using generic ValueError and adding more descriptive error messages.

- Maybe adding tests for concurrent session access, as well as adding more edge cases in error handling tests would be a good move.

AdithyahNair commented on src/components/ai_conversation_client/api.py:
The interface design is excellent - it properly implements dependency injection principles. The method signatures are well-defined with appropriate parameters and return types. The docstrings are particularly thorough, with clear Args/Returns/Raises sections that make the API easy to understand and use correctly.

AdithyahNair commented on src/components/ai_conversation_client/factory.py:
The factory implementation is clean and follows good design patterns. However, the docstrings for the create_client method and other functions could be more detailed. Consider adding examples of usage and more information about the expected parameters.

AdithyahNair commented on src/components/ai_conversation_client/tests/test_ai_client.py:
The test coverage is impressive, with tests for most functionality. To make it even more robust, consider adding tests for concurrent access patterns (multiple threads accessing the same session) and more edge cases around error conditions like network failures or invalid inputs.

Loke7132 committed:
TA and Peer Feedback

Loke7132 committed:
TA and Peer Feedback test-1

Loke7132 committed:
TA and Peer Feedback test-2

BhanuPrakashNani committed:
Merge branch 'interface-definition' into interface-implementation

BhanuPrakashNani commented:
> * Missing docstrings in factory.py and cli.py—please document all public functions and CLI commands to improve API discoverability.
> * Unit tests cover core interface and client logic, but add dedicated tests for CLI execution and error scenarios (e.g., invalid session IDs, network failures) to boost coverage.
> * Enhance cerebras_client.py with explicit parameter validation and robust error handling
> * Changes not relevant to implementation should not be present in this PR (things like changing test for other components, etc). Merge them separately if you require them in your repository, but having them in this implementation PR will only harm your grade.

Thank you for your thorough review! We’ve incorporated all your suggestions:

Added missing docstrings in factory.py and cli.py.
Expanded test coverage for CLI error scenarios (invalid session IDs, network failures).
Enhanced cerebras_client.py with parameter validation and error handling.

Regarding your note about “changes not relevant to implementation,” could you kindly clarify which specific changes you’d like us to remove or split into a separate PR? If you could point to one or two examples, we’ll address those immediately to keep this PR focused.

BhanuPrakashNani committed:
Update .coveragerc

BhanuPrakashNani committed:
Update .coveragerc

BhanuPrakashNani committed:
Update pyproject.toml

BhanuPrakashNani committed:
Update .coveragerc

BhanuPrakashNani committed:
Update .coveragerc

BhanuPrakashNani committed:
Update pyproject.toml
